docs/corfu-queue-design.md

```diff
@@ -0,0 +1,12 @@
+The class `CorfuQueue` implements a persisted queue over the abstraction of a CorfuTable. Currently CorfuTable uses a HashMap<> to represent the materialized view of a distributed map. A HashMap does not carry a notion of ordering natively. However when implemented over the abstraction of a distributed shared log, the elements added to the map do, in fact, have an ordering imposed by their append operations into the global log. The class `CorfuQueue` merely exposes this inherent ordering as an immutable logical FIFO Queue interface with three simple apis:
+
+###  1. `String enqueue(E)`
+Since a map has a key and value, where key is a conflict parameter, enqueue simply generates a non-conflicting UUID as the key and inserts the Entry as a value into a CorfuTable. The generated UUID is returned for the caller.
+
+### 2. `List<Map.Entry<String, Object>> entryList()`
+Returns a list of all the entries along with their ids, sorted by their realized stream addresses (thus logically a FIFO queue). Assuming insertion order is preserved, CorfuTable can materialize its state as a [LinkedHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html) instead of a HashMap and return the view as a list.
```

Comment:
What does this mean WRT undo records. Do we need to keep them around, or do we just remove the head of the list?

Reply:
If you mean the entryList() api then that is just a snapshot of a point in time. No undo records should be necessary.

Reply:
undo records are generated for mutations during syncing. They need to be generated for mutators, but there is  no explicit control over them, they are cached in the address space.